nwb_conversion_tools/utils/genericdatachunkiterator.py

```diff
@@ -0,0 +1,93 @@
+"""Authors: Cody Baker, Saksham Sharda, Oliver Ruebel."""
+from typing import Iterable, Tuple
+import numpy as np
+import psutil
+from abc import abstractmethod
+from hdmf.data_utils import AbstractDataChunkIterator, DataChunk
+from itertools import product
+
+
+class GenericDataChunkIterator(AbstractDataChunkIterator):
+    """DataChunkIterator that lets the user specify chunk shapes."""
+    
+    def _get_chunksize_mb(self, chunk_shape: Iterable, typesize: int) -> float:
+        return np.product(chunk_shape) * typesize / 1e6
+
+    def _set_chunk_shapes(self, chunk_shape: Iterable, typesize: int, buffer_mb: float) -> tuple:
+        n_dim = len(chunk_shape)
+        iter_idx = 0
+        while (self._get_chunksize_mb(chunk_shape, typesize) > buffer_mb) and np.product(chunk_shape) > 2:
+            dim = iter_idx % n_dim
+            chunk_shape[dim] = np.ceil(chunk_shape[dim] / 2.0)
+            iter_idx += 1
+        return tuple(int(x) for x in chunk_shape)
+
+    def __init__(self, chunk_shape: tuple = (), buffer_mb: float = 20.):
+        """
+        Initialize the SpecDataChunkIterator object with specified chunk parameters.
+
+        data : Iterable
+            The data to be chunked. Recommended to be a np.memmap at this point in time.
+        chunk_shape : tuple, optional
+            The desired shape of the chunks. Defaults to empty.
+        buffer_mb : int, optional
+            If chunk_shape is not specified, it will be inferred as the smallest chunk below the buffer_mb threshold.
+            Defaults to 20 MB.
+        """
+        assert buffer_mb > 0 and buffer_mb < psutil.virtual_memory().available / 1e6, \
+            f"Not enough memory in system handle buffer_mb of {buffer_mb}!"
+        self.data = self._get_data()
+        self.full_shape = self._get_shape()
+        self.__dtype = self._get_dtype()
+        if chunk_shape == ():
+            self.chunk_shape = self._set_chunk_shape(
+                chunk_shape=list(self.full_shape),
```

Comment:
why does `chunk_shape=full_shape`?

Reply:
More appropriately named to "initial_shape" since that is the role it plays in the iterative approach. Still working on analytic solution, will update docstrings when it's working better.